### PR TITLE
Generic SyncHeader type

### PR DESCRIPTION
**Pull request by svyatonik**
(to simplify integration with Kusama/Polkadot)

### CODE REVIEW

File: relays/substrate-client/src/sync_header.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+}
+
+impl<Header: HeaderT> SourceHeader<Header::Hash, Header::Number> for SyncHeader<Header> {
+	fn id(&self) -> HeaderId<Header::Hash, Header::Number> {
+		relay_utils::HeaderId(*self.number(), self.hash())
+	}
+
+	fn parent_id(&self) -> HeaderId<Header::Hash, Header::Number> {
+		relay_utils::HeaderId(*self.number() - One::one(), *self.parent_hash())
```
Comment (tomusdrw) [inline relays/substrate-client/src/sync_header.rs:54]: I guess we never use this with genesis header, right? I'd still prefer to be on the safe side here and use `saturating_sub`

File: relays/substrate-client/src/sync_header.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use sp_runtime::traits::Header as HeaderT;
+
+/// Generic wrapper for `sp_runtime::traits::Header` based headers, that
+/// implements `headers_relay::sync_types::SourceHeader` and may be used in headers sync directly.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SyncHeader<Header>(Header);
+
+impl<Header> SyncHeader<Header> {
+	/// Extracts wrapped header from self.
+	pub fn extract(self) -> Header {
```
Comment (tomusdrw) [inline relays/substrate-client/src/sync_header.rs:29]: I think the usual convention is `into_inner`, wdyt?

File: relays/substrate-client/src/sync_header.rs (hunk 1 above)
Comment (svyatonik) [reply relays/substrate-client/src/sync_header.rs:54]: I've added panic-related comment to the method definition. Swallowing error here may lead to hard-to-realize/find issues (hot loops, async hot loops), so panic vs Option/Result. Panic here is imo more appropriate (that's not the critical code).

Comment (tomusdrw) [reply relays/substrate-client/src/sync_header.rs:54]: Hmm, but it's only going to panic in debug mode, and overflow in `--release`. Maybe better to panic explicitly then?

Comment (svyatonik) [reply relays/substrate-client/src/sync_header.rs]: Right, I completely forgot that

**Review by tomusdrw — APPROVED** [review]

**Review by HCastano — APPROVED** [review]

**Pull request merged**